plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParserFsmNode.java

```diff
@@ -0,0 +1,135 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.parser;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.jkiss.dbeaver.parser.grammar.nfa.GrammarNfaOperation;
+
+/**
+ * Node of the parser finite state machine representing logical position in the text between some terminals
+ */
+class ParserFsmNode {
+    private final int id;
+    private final List<ParserFsmStep> steps = new ArrayList<>();
+    private final Map<String, TermGroup> stepsByTermGroup = new HashMap<>();
+    private final List<ParserFsmStep> finalSteps = new ArrayList<>();
+    private final boolean isEnd;
+    private Pattern pattern;
+
+    /**
+     * A bunch of transitions associated with the same terminal
+     */
+    private static class TermGroup {
+        private final String groupName;
+        private final String pattern;
+        private final List<ParserFsmStep> steps;
+
+        public TermGroup(String groupName, String pattern, List<ParserFsmStep> steps) {
+            this.groupName = groupName;
+            this.pattern = pattern;
+            this.steps = steps;
+        }
+    }
+
+    public ParserFsmNode(int id, boolean isEnd) {
+        this.id = id;
+        this.isEnd = isEnd;
+    }
+
+    public boolean isEnd() {
+        return this.isEnd;
+    }
+
+    public int getId() {
+        return this.id;
+    }
+
+    public Iterable<ParserFsmStep> getTransitions() {
+        return this.steps;
+    }
+
+    /**
+     * Register transition from the current node to the target node,
+     * associate it to the given patterns and grammar operations
+     * @param target node
+     * @param pattern
+     * @param operations
+     */
+    public void connectTo(ParserFsmNode target, String pattern, List<GrammarNfaOperation> operations) {
+        this.steps.add(new ParserFsmStep(this, target, pattern, operations));
+    }
+
+    /**
+     * Prepare common recognizing pattern for all possible terminals associated with presented parsing steps
+     */
+    public void prepare() {
+        Map<String, List<ParserFsmStep>> stepsByTerm = new HashMap<>();
+        for (ParserFsmStep s : this.steps) {
+            if (s.getPattern() != null) {
+                stepsByTerm.computeIfAbsent(s.getPattern(), p -> new ArrayList<>()).add(s);
+            } else {
+                finalSteps.add(s);
+            }
+        }
+
+        List<String> parts = new ArrayList<>(stepsByTerm.size());
+        for (Entry<String, List<ParserFsmStep>> step : stepsByTerm.entrySet()) {
```

Comment:
```suggestion
        for (var step : stepsByTerm.entrySet()) {
```

I believe it makes it a bit easier to read. It's a matter of preference, though.